### ISSUE TITLE
Unexpected changes in CSE

### ISSUE THREAD

**Issue by alex-mckenna** [clash-lang/clash-compiler#1846]
In f68204ff44bafdade476aa006dce0229aa5a2742, a mistake was fixed that was stopping Clash from identifying when an expression changed unexpectedly. Currently on `master`, we can see this in action by running

```bash
clash --vhdl -fclash-debug DebugApplied examples/Blinker.hs
```

which reports

```
<no location info>: error:                       
    Clash error call:                                                                                                                                                                                              
    Clash.Rewrite.Util(279): Expression changed without notice(CSE): beforeletrec
    ...
```

There is likely a path in CSE where a change is not reported, i.e. `return` is used instead of `changed`.
